### PR TITLE
feat(core-input): TDS-1128

### PR DESCRIPTION
**Pull request by ah-arch**
## Related issues

See #893 and #894, and [TDS-1128](https://telusdigital.atlassian.net/browse/TDS-1128).

## Description

- Adds `hintPosition` to align `hint` below the `label`
- Adds `node` to `error` propTypes to allow strings with `Link`s

## Checklist before submitting pull request

- [x] New code is unit tested
- [x] Commits follow our [Developer Guide](https://tds.telus.com/contributing/developer-guide.html#make-a-commit)
- [x] For code changes, run `yarn prepr` locally
  - make sure visual and accessibility tests pass

### CODE REVIEW

**Comment by marcod1419** [comment]
Hi @ah-arch! Just wanted to let you know that we see your pull request and we'll review it as soon as we are able. Thanks for contributing!

File: packages/Input/Input.md [hunk 1]
```diff
@@ -89,13 +89,17 @@ that the field is optional. It is a more usable and accessible option than the H
...
+<Input
+  label="Quick view"
+  hintPosition="below"
+  hint="Enter the account number in the box below to go to the overview page for that account."
+/>
 ```
+
+### Usage criteria
+
+- The `hint` should not be more than 85 characters when positioned below the label
```
Comment (simpleimpulse) [inline packages/Input/Input.md]: Consider adding more comprehensive usage criteria, recommending for what character lengths should be using 'inline' vs 'below'.

File: packages/Input/__tests__/Input.spec.jsx [hunk 2]
```diff
@@ -69,6 +69,20 @@ describe('Input', () => {
...
+    it('can have a long hint', () => {
+      const { input } = doMount({
+        id: 'the-id',
+        hint: 'The long hint over here',
+        hintPosition: 'below',
+      })
+
+      expect(input).toContainReact(
+        <Paragraph id="the-id_hint" size="medium">
+          The long hint over here
```
Comment (simpleimpulse) [inline packages/Input/__tests__/Input.spec.jsx]: Consider using an example of a hint that is so long that it can't be used as an inline hint.

Comment (ah-arch) [reply packages/Input/__tests__/Input.spec.jsx]: Will find 85 characters of lorem ipsum!

File: packages/Input/Input.md (hunk 1 above)
Comment (ah-arch) [reply packages/Input/Input.md]: Is there a character cut-off for using `inline` vs `below`? What should be recommended?

File: packages/Input/__tests__/Input.spec.jsx (hunk 2 above)
Comment (simpleimpulse) [reply packages/Input/__tests__/Input.spec.jsx]: You're right that this doesn't make any difference while programmatic testing, so we can safely skip this part.

File: packages/Input/Input.md (hunk 1 above)
Comment (ah-arch) [reply packages/Input/Input.md]: Updated based on conversation on TDS-1128

File: packages/Input/Input.jsx [hunk 3]
```diff
@@ -60,7 +64,7 @@ Input.propTypes = {
   /**
    * An error message. Either an error or a helper should be used, not both.
    */
-  error: PropTypes.string,
+  error: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
```
Comment (theetrain) [inline packages/Input/Input.jsx]: Let's change `PropTypes.node` to `componentWithName('Link')` to restrict this to text and the core-link component.

File: packages/Input/Input.md [hunk 4]
```diff
@@ -89,13 +89,20 @@ that the field is optional. It is a more usable and accessible option than the H
...
-)
-;<Input label="Credit Card Number" helper={creditCards} />
+<Input
+  label="Quick view"
+  hintPosition="below"
+  hint="Enter the account number in the box below to go to the overview page for that account."
+/>
 ```
+
+### Usage criteria
```
Comment (theetrain) [inline packages/Input/Input.md]: Please move the usage criteria section to the top of this document.

File: packages/Input/Input.md [hunk 5]
```diff
@@ -89,13 +89,20 @@ that the field is optional. It is a more usable and accessible option than the H
...
+
+- Use `hintPosition` to re-position the `hint` below the `label` if the inline `hint` wraps
+  - The `hint` should **not** be longer than 85 characters
+  - The `hint` should **not** wrap more than 3 lines on mobile
```
Comment (theetrain) [inline packages/Input/Input.md]: I'm not 100% sure here, but perhaps it should not wrap more than 3 lines `on all viewport sizes`

File: shared/components/FormField/FormField.jsx [hunk 6]
```diff
@@ -170,9 +185,10 @@ class FormField extends React.Component {
 FormField.propTypes = {
   label: PropTypes.string.isRequired,
   hint: PropTypes.string,
+  hintPosition: PropTypes.oneOf(['inline', 'below']),
   value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
   feedback: PropTypes.oneOf(['success', 'error']),
-  error: PropTypes.string,
+  error: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
```
Comment (theetrain) [inline shared/components/FormField/FormField.jsx]: Use `componentWithName` as mentioned above.

File: shared/components/FormField/__tests__/FormField.spec.jsx.md [hunk 7]
```diff
@@ -1 +1 @@
-// FormField is used by Textarea and all functionality is tested through it.
+// FormField is used by Textarea and Input, and all functionality is tested through them.
```
Comment (theetrain) [inline shared/components/FormField/__tests__/FormField.spec.jsx.md]: Good catch! I poked around and noticed `Select` also uses FormField.

Comment (ah-arch) [reply shared/components/FormField/__tests__/FormField.spec.jsx.md]: Will update to:
> FormField is used by Input, Textarea and Select, and all functionality is tested through them.

File: packages/Input/Input.md (hunk 5 above)
Comment (ah-arch) [reply packages/Input/Input.md]: Will update to:
>  The `hint` should **not** wrap more than 3 lines on all viewport sizes

File: packages/Input/Input.jsx (hunk 3 above)
Comment (ah-arch) [reply packages/Input/Input.jsx]: But we need to allow a combination of strings **and** Link: https://github.com/telus/tds-core/issues/894#issuecomment-455325223

Seems like [componentWithName](https://github.com/telus/tds-core/blob/367747116d1afcd7b3be304db84c7846093c267e/packages/prop-types/PropTypes.md#componentwithnamestring) would force the whole content to be a `Link`

This could be updated to use `or` though:

> `or([PropTypes.string, PropTypes.node])`

File: packages/Input/Input.md (hunk 4 above)
Comment (ah-arch) [reply packages/Input/Input.md]: Will move it below the `Minimal usage` section

File: shared/components/FormField/FormField.jsx (hunk 6 above)
Comment (ah-arch) [reply shared/components/FormField/FormField.jsx]: Leaving it due to reason mentioned [above](https://github.com/telus/tds-core/pull/920#discussion_r259396513).

**Comment by ah-arch** [comment]
> Changes:
>  - @tds/core-input: 2.0.0 => 2.1.0
>  - @tds/core-select: 2.0.0 => 2.0.1
>  - @tds/core-text-area: 2.0.0 => 2.0.1
>  - @tds/shared-form-field: 2.0.0 => 2.1.0 (private)

Seems like we are at `2.0.0` without deprecating the `helper`. Is this description still valid?
https://github.com/telus/tds-core/blob/8fa6173f750e7140cbd5ffc1c0d52a75d26c7e64/packages/Input/Input.jsx#L77-L78

File: packages/Input/Input.jsx [hunk 8]
```diff
@@ -70,6 +74,9 @@ Input.propTypes = {
    *
    * **Deprecation:** As of v1.0.0, passing `Input.Helper` will no longer work.
    *
+   * **Deprecation:** As of v2.0.0, it will no longer exist, as it is not a recommended pattern.
```
Comment (theetrain) [inline packages/Input/Input.jsx]: This appears to be a moving target on our part. We sunset features based on adoption, not version. Let's rephrase this to:
> This is not a recommended pattern and will be removed in a future release.

File: packages/Input/Input.jsx (hunk 3 above)
Comment (theetrain) [reply packages/Input/Input.jsx]: Can we use `componentWithName('Link')` here and in FormField?

Comment (ah-arch) [reply packages/Input/Input.jsx]: Ok, will update to:
> PropTypes.oneOfType([PropTypes.string, componentWithName('Link')])

Comment (ah-arch) [reply packages/Input/Input.jsx]: Updating to `error: or([PropTypes.string, componentWithName('Link')])` will cause the following warning:
error

When used with the following examples in https://github.com/telus/tds-core/issues/894, e.g.:
```<Input error={<React.Fragment>Hello <Link href="#">What is this</Link></React.Fragment>}  />```

Are we not allowing this format of messages?

File: packages/Input/Input.md [hunk 9]
```diff
@@ -14,12 +14,19 @@ For information on how to use disabled Inputs, please refer to the [disabled for
...
 </Box>
 ```
 
+### Usage criteria
+
+- Use `hintPosition` to re-position the `hint` below the `label` if the inline `hint` wraps
+  - The `hint` should **not** be longer than 85 characters
+  - The `hint` should **not** wrap more than 3 lines on all viewport sizes
+- We recommend `React.Fragment` when supplying custom strings to the `error` prop
+  - Note that the `div` tag is not valid in a `p` tag
```
Comment (theetrain) [inline packages/Input/Input.md]: Let's add some more notes here:

- Keep `error` text as brief as possible, should be limited to text and links

File: packages/Input/Input.jsx (hunk 3 above)
Comment (theetrain) [reply packages/Input/Input.jsx]: Please rewrite this as `error: PropTypes.node,` as we'll accept all children for now, combined with additional usage criteria to guide designers and developers on their use of the `error` prop.

**Comment by jraff** [comment]
Closing, see https://github.com/telus/tds-core/pull/952 for new PR.
